Localization with Fluent

Heinenen opened the pull request:
This pull request implements #216, some discussion took place there already.
Fluent website: https://projectfluent.org/

## Goals
This PR tries to integrate fluent into Typst.
It implements the most basic feature of retrieving a localized string for a given element (similar to a key-value store).
The immediate benefit that we gain from this PR is that the translations are no longer hard-coded and scattered across multiple source files. Instead, all translation (`.ftl`) files can reside in one singular location. This also simplifies the process of accepting new translations.

Note: For now, the translation files need to be present inside the project they are used in. What is the best way to include all the translation files inside the binary?

## Open questions
- (non-blocking) Do we want to provide an interface for users to provide their on translation files? The main downside I see with this, is that we lock ourselves into accepting a certain format, e.g., `.ftl` files.

edit: use correct file extension (`ftl`)

istudyatuni commented:
In sources you are using `.ftl`, but in PR description - `.flt`

LaurenzV commented on crates/typst/src/engine.rs:
```diff
@@ -40,6 +46,48 @@ impl Engine<'_> {
+            let mut bundle = FluentBundle::new(vec![langid_en]);
+            bundle
+                .add_resource(res)
+                .expect("Failed to add FTL resources to the bundle.");
```
Afaik all error messages in Typst are lower case. For consistency, you should probably do the same.

I'm also wondering whether such specific message should really be exposed to the user? Maybe just a generic "failed to load translations for specified language" would be enough? Not sure though.

alerque replied:
Once the resources are embeded in the binary (which they probably should be, as mentioned in the issue) this will / should be update to only possible throw a build time error, not a run time one anyway. For now this is only developer facing not user facing because it isn't a production arrangement yet (as I understood the PR state).

emilyyyylime replied:
Also this is an incorrect use of `expect`. `expect` isn't an assertion, see https://doc.rust-lang.org/std/result/enum.Result.html#recommended-message-style

alerque replied:
How is this incorrect? Isn't `.expect()` basically just "try to unwrap() and if success move on, otherwise throw this text string as an error if failed"?

Heinenen replied:
What I think they are saying is: It is correct to call `expect` here, but the wording of its argument is incorrect. It should be `expect("there needs to be a valid fluent file in the project root")` instead of `expect("no valid fluent file in project root")`.
I admit, I just copied the code from the example in https://docs.rs/fluent/0.16.0/fluent/ to have something working.

Heinenen replied:
The next thing I'm gonna do is migrating all languages to ftl files and including them into the binary.

LaurenzV commented:
Just a small note, I had thought about adding fluent to Typst before, what stopped me was mainly the fact that I didn't know whether this could be applied to enable localization in all areas of Typst. For example, It would in theory be nice if we could also use Fluent to provide the localization strings for different languages when formatting a date, but this would probably require us to somehow combine this with icu4x, and I don't know how compatible they are. Another issue would be things like how numbers are formatted by default in different languages, etc.

Not saying that this should be a blocker for adding Fluent to Typst, maybe those are separate problems and it's fine to use this just for the localized names! Just something to consider.

emilyyyylime commented:
I think it would be awesome to allow user-specified localisation files and more importantly—localisation entries. This would allow package authors to make their packages easily localised, which would be a huge advantage to using Typst!

alerque commented:
Fluent is moving to use icu4x internally anyway.

Heinenen commented:
> In sources you are using `.ftl`, but in PR description - `.flt`

Thanks, I edited it. `ftl` is correct.

Heinenen commented:
I ended up keeping most of the `expect` messages.
My reasoning being is that it's a logic bug in Typst and (if I implemented everything correctly) should never occur.
And if such a bug occurs, crashing the compiler instead of failing in a "softer" way seems reasonable to me.

Although it would be nicer to check the validity of the included `ftl` files at build time, I don't know if that is easily possible/worth the complexity (this would need to call the fluent parser at build time).

I implemented a simple fallback mechanism for the language strings:
- If no translation is present in the file for a language + region: search in language (without specified region)
- if no translation is present in the file for a language: search in English file 
- if no translation is present in English: logic bug -> panic

tingerrr commented:
> I think it would be awesome to allow user-specified localisation files and more importantly—localisation entries. This would allow package authors to make their packages easily localised, which would be a huge advantage to using Typst!

#2329 can serve as inspiration on how to expose this with a built-in element, although this would likely be an element that takes a fluent string and returns a function to accept arguments and such.

I'm excited to see proper localization come to typst.

laurmaedje commented:
To fluent in general: I would like to see some more motivation for using it. For what it is doing here (key -> value mapping), it is a pretty heavy dependency and it's runtime overhead also looks considerable. It's clearly better than the current match statements, but implementing that without fluent would also be quite simple. I understand that we might need more of fluent's capabilities in the future, but I would like to have a clearer picture of what those would be. Moreover, I am a bit concerned by the fact that fluent-rs practically didn't have commits in the last two years. It might very well be finished software, but it could also just be unmaintained. 

To this PR specifically: The language map shouldn't be stored in the engine. From my understanding, it is just a global storage for already loaded languages, all of which are embedded in the binary. Conceptually a static Lazy<HashMap> would do. Alternatively, we could even just use a memoized function that returns a FluentBundle (would then probably need to introduce an Arc because we couldn't return a reference, but the performance cost should be negligible, and we could also evict unused languages over time). I think we should also move as much of the cost into the function, so in particular conversion of Lang and Region to the kind of identifier fluent requires. The whole code could then be moved into `lang.rs` next to `LocalName`.

Heinenen commented:
> To fluent in general: I would like to see some more motivation for using it. For what it is doing here (key -> value mapping), it is a pretty heavy dependency and it's runtime overhead also looks considerable. It's clearly better than the current match statements, but implementing that without fluent would also be quite simple.

My plan would then be to write our own (very simple) parser for language files. See below for more details.
<br>
> Moreover, I am a bit concerned by the fact that fluent-rs practically didn't have commits in the last two years. It might very well be finished software, but it could also just be unmaintained.

My impression is that the basic building blocks are there, but a nice API around it is missing. On [crates.io](https://crates.io/crates/fluent#status) it says the following:
> > The implementation is in pre-1.0 mode and supports Fluent Syntax 1.0, and Fluent API 0.14.. Consult the [list of milestones](https://github.com/projectfluent/fluent-rs/milestones) for more information about release planning and scope.

Maybe @alerque can tell us more about that, he seems to be more involved with fluent.
<br>

Implementation: Your understanding of what I tried to do is correct. I expected some feedback of this kind, but I didn't know better.

Custom parser:
```rs
// pseudocode
let lines = file.lines();
for line in lines {
  let (key, val) = line.split_once('=');
  let (key, val) = (key.trim(), val.trim())
}
```
Should I change the `flt` file extension to something else to avoid confusion (and what would the new file extension be)?

tingerrr commented:
I think fluent does a little more than just key value replacements, perhaps we should look at the spec more closely before we jump to shipping a customized solution.

And if we go with a customized solution, we should likely think about the design a little more to allow this to be a good solution for user defined translations too.

PgBiel commented:
FWIW, `rust-i18n`[^1] uses simple YAML[^2] files which allow for some degree of substitution through `%{this format}`. I've used it in a project before, where (as a small detail) I made my own `tr!` wrapper around their `t!` macro so I could just specify e.g. `tr!(ctx=ctx, "key", value1="something")` instead of having to create a temporary variable to store the locale from the context and then doing `t!(locale=locale, "key", ...)`. Also, its error handling for when you try to specify a key or that doesn't exist or a locale without a translation for the key wasn't very ideal back then, but maybe it has improved by now (I think it now allows you to specify a fallback locale). Either way, that's something to consider; it was very simple and ergonomic to use, so it could work in the meantime.

[^1]: https://github.com/longbridgeapp/rust-i18n
[^2]: Although it seems JSON and TOML are also supported, which is great.

laurmaedje commented:
If we want to go with a simple custom key-value format, I'd be on board with this directly. If we want to go with fluent, I'd potentially also be on board, but would want to see a few more arguments in favor of it.

Heinenen commented:
I think I will then try to implement the custom key-value format.
I'm still new to this, should I then open a separate PR for the custom key-value format (and close this one)?

Either way, I think I want to finish up my other PR for named destinations first.

laurmaedje commented:
> I'm still new to this, should I then open a separate PR for the custom key-value format (and close this one)?

Depends on how much of a departure from the current PR you think it is. If you think the current state of the PR would be useful in the future should we switch to fluent, feel free to open a new one and close this one. Otherwise, you can also continue here. Up to you.

Heinenen commented:
Ok, then I will close this one.
